### ISSUE TITLE
Dynamic args for debug task

### ISSUE THREAD

**Issue by normalser** [Microsoft/vscode#9544]
It would be nice if one could define task with option `dynamicArgs: true` in `launch.json` (or something similar) that when starting debugger for the task - it would prompt user with text box to provide arguments that would be passed to the script (similar how current static `args` option works)

Use case: Ability to easily debug some cli tasks that can start with different arguments on command line

Right now one need to go to launch.json and edit arguments or duplicate configurations to handle different arguments

**Comment by weinand** [comment]
@wallverb mock-debug shows how this is already possible if you are an extension owner.
- The code for command that asks for a string: https://github.com/Microsoft/vscode-mock-debug/blob/master/src/extension.ts
- How to bind a new variable `AskForProgramName` to the command: https://github.com/Microsoft/vscode-mock-debug/blob/master/package.json#L60
- How to use the dynamic variable in a launch.json: https://github.com/Microsoft/vscode-mock-debug/blob/master/package.json#L86

**Comment by mike-pt** [comment]
So is this only possible if you're an extension owner?

I was just trying to use this for Python debugging.. but I get `command 'AskForProgramName' not found`

when I try to use:

```
    "configurations": [
        {
            "name": "Python",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "pythonPath": "${config.python.pythonPath}",
            "program": "${workspaceRoot}/${command.AskForProgramName}",
[ ..... ]
```

**Comment by weinand** [comment]
@mike-pt at the moment variables like `AskForProgramName` are only valid for the debugger extension that defined the variable.

However, it is already possible to use the command ID instead of the variable: `${command.extension.mock-debug.getProgramName}`.

Please note that you'll have to install the mock-debug extension for making the command `extension.mock-debug.getProgramName` available.

And we are planning to change the syntax a bit in the Februar milestone to this  `${command:extension.mock-debug.getProgramName}`

**Comment by mike-pt** [comment]
I see, thanks for the information.

**Comment by mike-pt** [comment]
Ok I was trying this but still got:
`command 'extension.mock-debug.getProgramName}' not found` same for askProgramName

I then realized that this is because I was still using "Python" as the debugger... so I deleted the launch.json and selected "mock debug", which works but it also means I can not specify a pythonpath... or args etc..

So still this is something targeted for extension owners not users...

You be nice if we could just naively do this...  for me it would be mostly for args, the program I'm debugging can use different args and It would be useful to be asked for which args to run... for now I'll just hardcode them and change as needed.

But thanks once more, at least now I understand how it works (I think :) )

**Comment by weinand** [comment]
@mike-pt variable substitution occurs **before** the launch config is passed to a debug adapter. But the issue is probably that the python  `args` expects an array and `${command.extension.mock-debug.getProgramName}` returns a string.

BTW, a `launch-config` can contain launch configurations for different debuggers, you don't have to "delete" them when switching debuggers.

**Comment by mike-pt** [comment]
Yeah that is the "issue" with arguments in the Python debugger... I don't see why its not a string though, Its not a bad idea to open an issue in their repo and ask for that.... 

With the proper changes we could then use "${command.extension.mock-debug.getProgramName}" in `args` and input the full args "line", but that's something to ask in the python debugger extension repo I suppose :)

I'm very recent to vscode (at least on a daily usage) so I'm not sue about codding my own extension, I can give it a try though.

Deleting was just easier to make sure it would offer the "mock-debug choice" when you don't have a launch-config at all.

**Comment by weinand** [comment]
Yes, the Python debugger could provide its own `${command.extension.python-debug.getCommandline}` (or even define a nice Variable `${command.AskProgramArguments}`).
And they could allow a single string for the `args` argument so that they could use the variable directly.

**Comment by rozzzly** [comment]
I've been following this thread, and was inspired to create an extension that helps prompt the user for a file to debug. It's a little rough around the edges currently, but I encourage you to check it out.

[rozzzly-dev.vscode-prompt-debug](https://marketplace.visualstudio.com/items?itemName=rozzzly-dev.vscode-prompt-debug)

![example](https://raw.githubusercontent.com/rozzzly/vscode-prompt-debug/master/img/anim.gif)

**Comment by weinand** [comment]
Closing, since the existence of the vscode-prompt-debug extension proves that dynamic attributes are possible.

**Comment by weinand** [comment]
This feature will become available in the November release via #63987. 

Example:
```ts
{
	"version": "0.2.0",
	"configurations": [
		{
			"type": "node",
			"request": "launch",
			"name": "Launch Program",
			"program": "${workspaceRoot}/${input:programName}",
		}
	],
	"inputs": [
		{
			"id": "programName",
			"type": "pickString",
			"default": "test.js",
			"description": "The program name"
		}
	]
}
```
